Added support for classes as a job parameter.

Muccarini opened the pull request:
Extended DisplayDataAttribute to Properties

![Screenshot 2025-03-14 120802](https://github.com/user-attachments/assets/13d508f9-6a35-4f7e-9da1-1cd55d9d0ec6)

Muccarini committed:
Added support for custom classes in automatic input generation.

-Extend DisplayDataAttribute usage on properties
-Added methods for processing nested custom classes in ManagementBasePage and JobPartial
-Added HashSet<Type> to track processed nested types and prevent infinite recursion (Circular Ref).

Muccarini committed:
Added custom class handling in Expedited job example.

Muccarini commented:
@lcourson 
any updates on this? using more complex object as job parameter could be useful and time saver.

brodrigz commented:
Excellent addition, I've been using it myself and it makes the code a lot tidier, as a suggestion, do you think I'd be possible to allow inheritance in the class parameter?

for example, I'd have class X as parameter, and have classes that inherit it decorated with DisplayData, and in the UI side there could be a dropdown to select the class, making the selected class's panel visible.

*Edit: Also Interface support would be pretty cool!

Muccarini committed:
Added support for interfaces as a job parameter

A dropdown menu allows selection of the concrete implementation of the interface, enabling polymorphic job configurations.

Muccarini committed:
Added Job2 example of interfaces usage

Muccarini commented:
Thank you so much, @brodrigz , for your feedback, your suggestion, and your positive comment.

I’ve implemented support for interfaces as parameters, allowing the user to choose the desired concrete implementation through the UI, in order to leverage interface polymorphism.
For base classes, I would recommend overloading the jobs with the different child classes instead.

![image](https://github.com/user-attachments/assets/4c3d0ac5-2ee1-4c3c-8c07-3a7866686dc5)

brodrigz commented:
Excellent @Muccarini , this covers my exact use case, thank you for your time.

Btw, I've noticed a bug with how implementations are registered on `VT`, it seems that it won't look for nested interface implementations.

The `Job2 ` example only works because the nested interface `IInterfaceTest ` is the same type passed in the job parameter, therefore its registered already, if we try with a interface that is not passed on the job parameter, it won't find it:
![image](https://github.com/user-attachments/assets/e42bde7e-e085-4010-ad86-804efb14b1c2)
![image](https://github.com/user-attachments/assets/d02a8892-2bf2-4b41-8155-8a0f13ebc97a)

You could recursively look for interfaces while still avoiding circular references with something like this:
```
	public static class VT
	{
		public static Dictionary<Type, HashSet<Type>> Implementations { get; private set; } = new Dictionary<Type, HashSet<Type>>();

		internal static void SetAllImplementations(Assembly assembly)
		{
			// this gets impls from param interfaces, but not from nested interfaces
			//JobsHelper.Metadata.ForEach(job => job.MethodInfo.GetParameters().ToList().ForEach(param => {
			//	assembly.GetTypes()
			//		.Where(t => t.IsClass && !t.IsAbstract && t.GetInterfaces().Contains(param.ParameterType))
			//		.ToList()
			//		.ForEach(implType => {
			//			if (!Implementations.ContainsKey(param.ParameterType))
			//			{
			//				Implementations[param.ParameterType] = new HashSet<Type>();
			//			}
			//			Implementations[param.ParameterType].Add(implType);
			//		});
			//}));

			JobsHelper.Metadata.ForEach(job => job.MethodInfo.GetParameters().ToList().ForEach(param => RegisterInterfaceImpls(assembly, param.ParameterType)));
		}
		private static void RegisterInterfaceImpls(Assembly assembly, Type interfaceType)
		{
			// this is only needed for interfaces
			if (!interfaceType.IsInterface) return;

			// init collection if null
			if (!Implementations.ContainsKey(interfaceType))
			{
				Implementations[interfaceType] = new HashSet<Type>();
			}

			RegisterImplementations(assembly, interfaceType);
		}

		// recursive
		private static void RegisterImplementations(Assembly assembly, Type interfaceType)
		{
			var dictList = Implementations[interfaceType];

			// get impls
			var implementations = GetInterfaceImplementations(assembly, interfaceType).ToList();

			foreach (var impl in implementations)
			{
				// register impl
				if (!dictList.Contains(impl)) dictList.Add(impl);

				// get nested interfaces
				var nestedInterfaces = GetInterfacePropsFromType(impl)
					.Where(i => i != interfaceType) // avoids circular refferences
					.ToList();

				// register nested interfaces
				foreach (var nestedInterface in nestedInterfaces) RegisterInterfaceImpls(assembly, nestedInterface);
			}
		}

		// gets all properties which are interfaces inside a type
		private static IEnumerable<Type> GetInterfacePropsFromType(Type classType)
		{
			var types = classType.GetProperties().Select(p => p.PropertyType).ToList();

			return types.Where(t => t.IsInterface);
		}

		// gets all concrete impls of given interface
		private static IEnumerable<Type> GetInterfaceImplementations(Assembly assembly, Type interfaceType) => assembly.GetTypes().Where(t => interfaceType.IsAssignableFrom(t) && !t.IsInterface);
	}
```
I didn't test it much but it's working for my use cases, feel free to use any of it, and thanks again for your time.

Muccarini committed:
fixed implementations finder

Muccarini committed:
Merge pull request #1 from Muccarini/Implementations_fix

fixed implementations finder

Muccarini committed:
Update README.md

Muccarini committed:
Added: list as a parameter, nested property list also list of list of ... are supported.

Muccarini committed:
Merge pull request #2 from Muccarini/list-as-input

Added: list as a parameter, nested property list also list of list of…
